frame/nomination-pools/src/lib.rs

```diff
@@ -0,0 +1,1967 @@
+	/// respected.
+	fn try_inc_delegators(&mut self) -> Result<(), DispatchError> {
+		ensure!(
+			MaxDelegatorsPerPool::<T>::get()
+				.map_or(true, |max_per_pool| self.delegator_counter < max_per_pool),
+			Error::<T>::MaxDelegators
+		);
+		ensure!(
+			MaxDelegators::<T>::get().map_or(true, |max| Delegators::<T>::count() < max),
+			Error::<T>::MaxDelegators
+		);
+		self.delegator_counter = self.delegator_counter.defensive_saturating_add(1);
+		Ok(())
+	}
+
+	/// Decrement the delegator counter.
+	fn dec_delegators(mut self) -> Self {
+		self.delegator_counter = self.delegator_counter.defensive_saturating_sub(1);
+		self
+	}
+
+	/// The pools balance that is transferrable.
+	fn transferrable_balance(&self) -> BalanceOf<T> {
+		let account = self.bonded_account();
+		T::Currency::free_balance(&account)
+			.saturating_sub(T::StakingInterface::active_stake(&account).unwrap_or_default())
+	}
+
+	fn can_nominate(&self, who: &T::AccountId) -> bool {
+		*who == self.roles.root || *who == self.roles.nominator
+	}
+
+	fn can_kick(&self, who: &T::AccountId) -> bool {
+		*who == self.roles.root ||
+			*who == self.roles.state_toggler && self.state == PoolState::Blocked
+	}
+
+	fn can_toggle_state(&self, who: &T::AccountId) -> bool {
+		*who == self.roles.root || *who == self.roles.state_toggler && !self.is_destroying()
+	}
+
+	fn can_set_metadata(&self, who: &T::AccountId) -> bool {
+		*who == self.roles.root || *who == self.roles.state_toggler
+	}
+
+	fn is_destroying(&self) -> bool {
+		matches!(self.state, PoolState::Destroying)
+	}
+
+	/// Whether or not the pool is ok to be in `PoolSate::Open`. If this returns an `Err`, then the
+	/// pool is unrecoverable and should be in the destroying state.
+	fn ok_to_be_open(&self) -> Result<(), DispatchError> {
+		ensure!(!self.is_destroying(), Error::<T>::CanNotChangeState);
+
+		let bonded_balance =
+			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+		ensure!(!bonded_balance.is_zero(), Error::<T>::OverflowRisk);
+
+		let points_to_balance_ratio_floor = self
+			.points
+			// We checked for zero above
+			.div(bonded_balance);
+
+		// Pool points can inflate relative to balance, but only if the pool is slashed.
+		// If we cap the ratio of points:balance so one cannot join a pool that has been slashed
+		// 90%,
+		ensure!(points_to_balance_ratio_floor < 10u32.into(), Error::<T>::OverflowRisk);
+		// while restricting the balance to 1/10th of max total issuance,
+		ensure!(
+			bonded_balance < BalanceOf::<T>::max_value().div(10u32.into()),
+			Error::<T>::OverflowRisk
+		);
+		// then we can be decently confident the bonding pool points will not overflow
+		// `BalanceOf<T>`. Note that these are just heuristics.
+
+		Ok(())
+	}
+
+	/// Check that the pool can accept a member with `new_funds`.
+	fn ok_to_join(&self) -> Result<(), DispatchError> {
+		ensure!(self.state == PoolState::Open, Error::<T>::NotOpen);
+		self.ok_to_be_open()?;
```

Comment:
is this defensive? if pool state is open, then it must have been checked before being set to `PoolSate::Open`.

Reply:
There is the case though that the pool is open but no one has poked it into closed